### PR TITLE
LAB-1342 resubmit experiment

### PR DESCRIPTION
**Pull request by supraja-968**
## What type of PR is this? 

- 🎮 Feature

## Description

backend logic for flow level aggregation.
this is a prerequisite for the unified view frontend updates. this PR only focuses on the backend logic
this PR also contains changes for [LAB-1408](https://linear.app/convexitylabs/issue/LAB-1408/seq-only-tool-updates)

## Related Tickets & Documents

Unified view linear ticket for reference: [LAB-1400](https://linear.app/convexitylabs/issue/LAB-1400/experiment-inputs-outputs-unified-view)
labsay seq-only tool updates are done in the Convexity repo in this PR: [#30](https://github.com/convexitylabs/convexity/pull/30)

Note: Combinatorial submissions are enabled to be able to test this feature

## Steps to Test

Use tool `labsay seq-only with checkpoints v0.4`
Submit 2 or more jobs combinatorially
Upload pdb files for checkpointing
The tool will generate random numbers for plddt and ipae scores for x and y axis

Expected output: Aggregated checkpoints (3 points per job) all shown together in the scatterplot. Conditions shown below the plot with their respective logs and output files.

### CODE REVIEW

File: frontend/app/experiments/ExperimentDetail.tsx [hunk 1]
```diff
@@ -156,7 +157,8 @@ export default function ExperimentDetail({ experimentID }: { experimentID: strin
               </div>
             </CardContent>
           </Card>
-
+          {/* add if checkpoint compatible later!! */}
```
Comment (thetechnocrat-dev) [inline frontend/app/experiments/ExperimentDetail.tsx:160]: Did you want to do this in this PR or a later PR?

File: gateway/handlers/checkpoints.go [hunk 2]
```diff
@@ -86,15 +155,34 @@ func ListCheckpointsHandler(db *gorm.DB) http.HandlerFunc {
 	}
 }
 
-func AggregateCheckpointData(jobUUID string) ([]models.ScatterPlotData, error) {
+// AggregateCheckpointData function aggregates the checkpoint data for a given flow. this could be a for loop that calls aggregateJobCheckpointData for each job in the flow. this function is on the flow level. this function is only passed the flowUUID and not the jobUUID. with that flowuuid, first find all jobs under it. then use each job to call aggregateJobCheckpointData. use getflowhandler to get all the jobs under the flow. then use the jobUUID to call aggregateJobCheckpointData. then return the aggregated data.
```
Comment (thetechnocrat-dev) [inline gateway/handlers/checkpoints.go:158]: I personally think the code is clear enough to not need a detailed comment description

File: gateway/server/server.go [hunk 3]
```diff
@@ -53,8 +53,10 @@ func NewServer(db *gorm.DB) *mux.Router {
 	router.HandleFunc("/datafiles/{cid}/download", protected(handlers.DownloadDataFileHandler(db))).Methods("GET")
 	router.HandleFunc("/datafiles", protected(handlers.ListDataFilesHandler(db))).Methods("GET")
 
-	router.HandleFunc("/checkpoints/{jobID}", handlers.ListCheckpointsHandler(db)).Methods("GET")
-	router.HandleFunc("/checkpoints/{jobID}/get-data", handlers.GetCheckpointDataHandler(db)).Methods("GET")
+	router.HandleFunc("/checkpoints/{flowID}", handlers.ListFlowCheckpointsHandler(db)).Methods("GET")
```
Comment (thetechnocrat-dev) [inline gateway/server/server.go:56]: Should/can these routes be wrapped with the `protected` middleware?

**Review by thetechnocrat-dev — APPROVED** [review]
Code looks good 💯 . Did not test it locally though, will do that if there is time after looking at the other PR. But, I saw the demo during retro.

Only comment that needs addressed is the `protected` route question.

File: frontend/app/experiments/ExperimentDetail.tsx (hunk 1 above)
Comment (supraja-968) [reply frontend/app/experiments/ExperimentDetail.tsx:160]: later PR, Jeff is working on the unified view. so i didnt concentrate on the frontend changes much. I can add this to the list of frontend updates

File: gateway/handlers/checkpoints.go (hunk 2 above)
Comment (supraja-968) [reply gateway/handlers/checkpoints.go]: agreed. I added the comment to get help from github copilot 😄  removed now

File: gateway/server/server.go (hunk 3 above)
Comment (supraja-968) [reply gateway/server/server.go:56]: yes, will raise a following PR for these changes

**Pull request merged**